controllers/propagator/template_utils.go

```diff
@@ -0,0 +1,645 @@
+	client, err := kubernetes.NewForConfig(kubeconfig)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	dynamicWatcherReconciler, dynamicWatcherSource := k8sdepwatches.NewControllerRuntimeSource()
+
+	return &TemplateResolvers{
+		contollerCtx:             ctx,
+		client:                   client,
+		config:                   kubeconfig,
+		defaultTemplateResolver:  defaultTemplateResolver,
+		dynamicWatcherReconciler: dynamicWatcherReconciler,
+		mgrClient:                mgrClient,
+		replicatedPolicyUpdates:  replicatedPolicyUpdates,
+		TemplateResolvers:        map[types.NamespacedName]*templateResolverWithCancel{},
+		wg:                       &sync.WaitGroup{},
+	}, dynamicWatcherSource, nil
+}
+
+// WaitForShutdown waits for the context passed to NewTemplateResolvers to complete and then for all goroutines
+// started from TemplateResolvers to end.
+func (t *TemplateResolvers) WaitForShutdown() {
+	<-t.contollerCtx.Done()
+	t.wg.Wait()
+}
+
+// initResolver will instantiate a DynamicWatcher with a token from the input service account and then instantiate a
+// TemplateResolver with that DynamicWatcher. Tokens will be automatically refreshed.
+func (t *TemplateResolvers) initResolver(
+	ctx context.Context,
+	serviceAccount types.NamespacedName,
+) (*templateResolverWithCancel, error) {
+	resolverCtx, resolverCtxCancel := context.WithCancel(ctx)
+
+	// Refresh the token anywhere between 1h50m and 1h55m after issuance.
+	tokenConfig := TokenRefreshConfig{
+		ExpirationSeconds: 7200, // 2 hours,
+		MinRefreshMins:    5,
+		MaxRefreshMins:    10,
+		OnFailedRefresh: func(originalErr error) {
+			if err := t.onBackgroundError(ctx, serviceAccount); err != nil {
+				log.Error(err, "Failed to handle the token refresh unexpectedly stopping", "originalErr", originalErr)
+			}
+		},
+	}
+
+	// Use BearerTokenFile since it allows the token to be renewed without restarting the template resolver.
+	tokenFile, err := GetToken(resolverCtx, t.wg, t.client, serviceAccount, tokenConfig)
+	if err != nil {
+		// Make this error identifiable so that up in the call stack, a watch can be add to the service account
+		// to trigger a reconcile when the service account becomes available.
+		if k8serrors.IsNotFound(err) {
+			err = errors.Join(ErrSAMissing, err)
+		}
+
+		resolverCtxCancel()
+
+		return nil, err
+	}
+
+	saConfig := getUserKubeConfig(t.config, tokenFile)
+
+	dynamicWatcher, err := k8sdepwatches.New(
+		saConfig,
+		t.dynamicWatcherReconciler,
+		&k8sdepwatches.Options{
+			DisableInitialReconcile: true,
+			EnableCache:             true,
+		},
+	)
+	if err != nil {
+		resolverCtxCancel()
+
+		return nil, fmt.Errorf(
+			"failed to instantiate the template resolver for the service account %s: %w", serviceAccount, err,
+		)
+	}
+
+	rv := templateResolverWithCancel{
+		cancel:         resolverCtxCancel,
+		referenceCount: &atomic.Uint32{},
+	}
+
+	t.wg.Add(1)
+
+	go func() {
+		err := dynamicWatcher.Start(resolverCtx)
+
+		rv.canceled.Store(true)
```

Comment:
Why this is located before `err != nil`?

Reply:
`Start` is blocking so if the method returns, it means the dynamic watcher is no longer running regardless of if stopping it was an error or intentional.

Reply:
I added a comment to clarify this.